### ISSUE TITLE
HistGradientBooster vs. GradientBooster

### ISSUE THREAD

**Issue by simpsus** [jefferythewind/scikit-learn-erasplit#6]
Is there a reason you chose to fork the Hist version of the Booster?
The hist version has a mandatory binning step which is completely unnecessary for numerai data and that step takes ~20 minutes on my setup ...

**Comment by jefferythewind** [comment]
Yea the Hist version is similar to the default implementation of LGBM, which we were going for. While the binning step takes time it then speeds up tree building substantially, which is why it was used in the pre-era split setting. In fact the era splitting work is implemented in the gradient boosting regressor as well, and also the decision tree regressor. I just didn't document that. These could be expanded upon as well.

**Comment by jefferythewind** [comment]
Without binning you need to evaluate the split criterion at every unique feature value when evaluating that feature. So I guess the binning doesn't save operations with the numerai data since it comes pre-binned. We could definitely look into the improving the gradient boosting regressor implementation. Thanks for the heads up on that. Just in the most general setting where features are not pre-binned, binning saves more time during tree-building than it takes to create the bins in the first place. But in era splitting you also need to create a bin for every era. In this way era splitting increases the time complexity of this algorithm by a factor of the number of eras in the data. In my experience it may help to group every N eras together. So example: data['era'] = round( data['era']/ N ). But yea you bring up a good point that maybe we maybe not gain anything binning the Numerai data.

**Comment by simpsus** [comment]
Maybe with daily rounds we have the potential to drastically reduce the number of eras by investigating which eras are similar.
Of course this ultimately leads to the topic of online-learning which era a new era is most similar to. I do not want go down that road here, here we want to group training data into bigger bins to make the training process more efficient (at least) or more effective (hopefully).

In terms of priority the binning is a nuisance and era-grouping is a potential improvement, but the possible identation error mentioned in #4 is a crucial question on the potential of the whole approach

**Comment by jefferythewind** [comment]
Yes the indentation error an important thing, like I said previously I have done several implementations of this in different code bases including the decision tree regressor and gradient boosting regressor as well as a completely different library, I think I introduced this indentation error when I had started adding in more additional criteria in addition to the era splitting criterion. I can find the commit that did that.
this line definitely needs to be inside the loop which loops over the data in each era. In each era we have to add an element to the numerator and denominator to be used in the Boltzmann operator.

thanks a lot for your attention, it's always good to get additional eyes on any piece of code. I'd love to mock up some simple runnable tests on the numerai data and put them in the repo. once I get a chance I will also try to share the other way I'm doing it in the decision tree regressor and the gradient boosting regressor.

**Comment by simpsus** [comment]
Do you know if the warm_start attribute of the constructur prevents the binning?
Anyway, I will find that out by testing and close this issue. Thank you very much for your work and help!

**Comment by jefferythewind** [comment]
warm_start means that you're continuing training and not starting from zero. For this to work you must have called 'fit' previously on the model. Example if you already trained to 1000 trees with warm_start == True, then you can update the max_iter param to 2000, calling fit will then continue fitting from 1000 to 2000. However I think it will still rebuild the histogram. Because with this option you could choose to build using new/different data on the second call to 'fit'.

**Comment by simpsus** [comment]
so warm_start does its job in the sense that a subsequent "fit" call runs very fast, but I am afraid it is no good as I am updating the target in my study and that warrants new splits. Those are however not updated with warm_start. I would assume warm start makes a lot of sense when you have the same data in features and labels, not just in features as I have (and is used for binning). So the binning step remains unnecessary but well ...

**Comment by jefferythewind** [comment]
Right, yes warm start is good if training takes a long time. The way I used it in the past is to save the model at checkpoints during training, like after every 1000 trees, but then yes I think you have to rebuild the histogram each time you call fit. Later on I found it adds too much extra time and I try to train the whole thing at once now.
